### PR TITLE
Allow parsing of dictionaries in constraint section

### PR DESCRIPTION
**Pull request by roberteg16**
This PR also introduces the changes to allow Constraint to have no inputs. Which means that now, the constraint is only under the precedence of the root position.

### CODE REVIEW

File: mlir/include/mlir/Dialect/PDL/IR/PDLOps.td [hunk 1]
```diff
@@ -49,7 +49,9 @@ def PDL_ApplyNativeConstraintOp
                        Variadic<PDL_AnyType>:$args,
                        DefaultValuedAttr<BoolAttr, "false">:$isNegated);
   let results = (outs Variadic<PDL_AnyType>:$results);
-  let assemblyFormat = "$name `(` $args `:` type($args) `)` (`:`  type($results)^ )? attr-dict";
+  let assemblyFormat = [{
+    $name (`(` $args^ `:` type($args) `)`)? (`:` type($results)^)? attr-dict
```
Comment (flemairen6) [inline mlir/include/mlir/Dialect/PDL/IR/PDLOps.td:53]: Why did this change? Does it mean that the arguments can be optional (a constraint without argument can be expressed without ()?

**Review by flemairen6 — APPROVED** [review]
LGTM, I'm just not sure why this PR enables optional arguments on constraints

**Comment by roberteg16** [comment]
> LGTM, I'm just not sure why this PR enables optional arguments on constraints

The reason is that we could have constraints that do have inputs, like `__builtin_createDictionaryAttr_constraint` which only returns a empty dictionary. Either we allow optional arguments or I insert dummy/fake arguments to the constraint

File: mlir/include/mlir/Dialect/PDL/IR/PDLOps.td (hunk 1 above)
Comment (roberteg16) [reply mlir/include/mlir/Dialect/PDL/IR/PDLOps.td:53]: Exactly that

**Pull request merged**